fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java

```diff
@@ -0,0 +1,205 @@
+        cloudConfiguration = connector.getMetadata().getCloudConfiguration();
+        Preconditions.checkState(cloudConfiguration != null,
+                String.format("cloudConfiguration of catalog %s should not be null", catalog));
+    }
+
+    public void setupScanRangeLocations(TupleDescriptor tupleDescriptor, ScalarOperator predicate,
+                                        List<PartitionKey> partitionKeys) {
+        List<String> fieldNames =
+                tupleDescriptor.getSlots().stream().map(s -> s.getColumn().getName()).collect(Collectors.toList());
+        List<RemoteFileInfo> fileInfos = GlobalStateMgr.getCurrentState().getMetadataMgr().getRemoteFileInfos(
+                table.getCatalogName(), table, partitionKeys, -1, predicate, fieldNames, -1);
+        RemoteFileDesc remoteFileDesc = fileInfos.get(0).getFiles().get(0);
+        OdpsSplitsInfo splitsInfo = remoteFileDesc.getOdpsSplitsInfo();
+        if (splitsInfo.isEmpty()) {
+            LOG.warn("There is no odps splits on {}.{} and predicate: [{}]",
+                    table.getDbName(), table.getTableName(), predicate);
+            return;
+        }
+        Map<String, String> commonSplitInfo = new HashMap<>();
+        String serializeSession = splitsInfo.getSerializeSession();
+        commonSplitInfo.put("read_session", serializeSession);
+        commonSplitInfo.putAll(splitsInfo.getProperties());
+        commonSplitInfo.put("split_policy", splitsInfo.getSplitPolicy().name().toLowerCase());
+        for (InputSplit inputSplit : splitsInfo.getSplits()) {
+            TScanRangeLocations scanRangeLocations = new TScanRangeLocations();
+            THdfsScanRange hdfsScanRange = new THdfsScanRange();
+            Map<String, String> splitInfo = new HashMap<>(commonSplitInfo);
+            splitInfo.put("session_id", inputSplit.getSessionId());
+            switch (splitsInfo.getSplitPolicy()) {
+                case SIZE:
+                    IndexedInputSplit split = (IndexedInputSplit) inputSplit;
+                    splitInfo.put("split_index", String.valueOf(split.getSplitIndex()));
+                    break;
+                case ROW_OFFSET:
+                    RowRangeInputSplit split1 = (RowRangeInputSplit) inputSplit;
+                    splitInfo.put("start_index", String.valueOf(split1.getRowRange().getStartIndex()));
+                    splitInfo.put("num_record", String.valueOf(split1.getRowRange().getNumRecord()));
+                    break;
+                default:
+                    throw new StarRocksConnectorException(
+                            "unsupported split policy: " + splitsInfo.getSplitPolicy().name());
+            }
+            hdfsScanRange.setOdps_split_info(splitInfo);
+            hdfsScanRange.setUse_odps_jni_reader(true);
+            hdfsScanRange.setFile_length(1);
+            // backend selector use length to load balancing
+            hdfsScanRange.setLength(1);
```

Comment:
Do we have plans to use native to support reading odps data?

Reply:
ODPS has standardized the interface to the complex storage layer using the Arrow format, making the Catalog reading logic almost native. Future plans may involve replacing JNI with a C++-based reading solution and adding the capability of index pushdown. There is currently no plan for direct filesystem reads.